BACKENDS: SAVES: WIN32: Use DeleteFile() instead of _tremove()

carlo-bramini opened the pull request:
I suggest to use `DeleteFile()` instead of `_tremove()` inside `backends\saves\windows\windows-saves.cpp`.
Both functions accept `TCHAR` as parameter, so they can be exchanged easily.
For reference, this is the page about `DeleteFile()` from MSDN:

https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-deletefile

Interesting parts of the above page are:
```
BOOL DeleteFile(
  [in] LPCTSTR lpFileName
);
```
```
If the function succeeds, the return value is nonzero.
If the function fails, the return value is zero (0).
To get extended error information, call GetLastError.
```
```
If an application attempts to delete a file that does not exist, the DeleteFile function fails with ERROR_FILE_NOT_FOUND.
If the file is a read-only file, the function fails with ERROR_ACCESS_DENIED.
```
Since this `windows-saves.cpp` is expected to be used only on platforms running Microsoft Windows, there is no need to keep the overhead introduced by `_tremove()`.

bluegr commented:
The main difference is that `remove()` and `_tremove()` are parts of the C runtime library:
https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/remove-wremove?view=msvc-170

I have no strong opinion on `DeleteFile()` vs `remove()`.

Has this been tested in older Windows versions? (i.e. XP and 98)

sluicebox commented:
Is there a technical reason to do this? I'm not aware of one, but I'm interested. The overhead of a CRT function in the call stack does not meet that threshold for me =)

If there is no technical reason to call one API over the other, then this is an arbitrary code change. In that case it's about code organization and maintainability. As a maintainer, I have thoughts.

Right now we have two near-identical implementations of `removeFile`. The only difference is the necessary Windows path conversion. Any developer can see that they are the same at a glance, even if they're not familiar with Windows-specifics, which is a sizable number of team members if not the majority. But with this change the two would be completely different. One would require Windows knowledge, and also now we're now responsible for two completely different error code translations.

We have two small functions that do the same thing, so they're written the same. This change would give us two small functions that are completely different, introduce another form of error code translation in addition to our proven one, and for all that... the best case scenario is they continue to do the exact same thing.

I don't think we should be accepting arbitrary changes to working API calls. If this is accepted, and a year later there is a PR titled "Replace DeleteFile with _tremove for consistency", why wouldn't we accept that? What would be the argument against it? This is what I mean by arbitrary.

There is probably no consequence to merging this PR. But if there really is no consequence, then there's no justification for making this change, and so we shouldn't. Am I missing something?

bluegr commented:
I agree with @sluicebox that now the code implemented in different backends is quite similar, and easy to follow. We should avoid platform-specific functions, when possible, which would affect the logic of the code, if we can help it. Also, there's no real point in choosing one function over the other, they are functionally the same.

Closing

carlo-bramini commented:
> Are you basing your port on our old port by Jubanka? https://github.com/scummvm/scummvm/tree/branch-2-1-0/backends/platform/wince

I started my port from scratch, by using current sources of SCUMMVM long time ago and updating them sometimes.
I didn't know about the existance of that work, actually.

carlo-bramini commented:
> Also, to address your PR in general, you could also subclass `WindowsSaveFileManager` and have this method overloaded in your port.

Thank you for your reply.

I have to say that I was hoping to use existing code of SCUMMVM without adding or changing (almost) anything, if possible.
Besides subclassing, it is also true that I can add an half-fake `remove()` function into `libmingw32` (see below) but unfortunately a fake `errno` cannot be really safe and I wanted to avoid it, if possible...

At the moment, my port uses serveral tools:

**The build environment:**
I'm using CYGWIN as development platform.
This means that an user must install it with usual development tools (like GNU make, for example), if they want to build SCUMMVM from sources.

**The cross compiler:**
I started from the excellent work made here:
https://github.com/MaxKellermann/cegcc-build
I updated binutils to version 2.42 and I added some stuff for building an `arm-w64-mingw32ce` cross compiler.
I did some additions also to the mingw runtime for adding functions typically not available on WinCE, including some Windows API missing here but available on Desktop versions of the Operating System.
This is the way I was using to ensure there were no changes to the code of SCUMMVM.
I would like to make these tools available as packages to be installed into an existing CYGWIN installation at my repository.

**Thirdy party libraries.**
I compiled as packages also some libraries for the cross compiler, but not all are used by SCUMMMVM:
- bzip2
- freetype2
- giflib
- libjpeg
- libpng
- libtiff
- mpg123
- ogg
- sdl-1.2
- sdl_image-1.2
- sdl_mixer-1.2
- sdl_net-1.2
- sdl_rtf-1.2
- sdl_ttf-1.2
- zlib

In the past, I sent patches to some of these projects for making them working (again) on WinCE, or for creating the conditions able to build them.

The version of sdl-1.2 needed to build for WinCE is 1.2.6git, because it includes those fixes but it won't be officially released anymore to the public.

It was my intention to build SCUMMVM for WinCE by using `--enable-plugins --default-dynamic` options and load engines only if needed. Unfortunately, nowadays this thing doesn't work yet.

For example, if you configure SCUMMVM with `--disable-all-engines --enable-engine=scumm` with `x86_64-w64-mingw32`, at the end of the process you will get two important files: `scummvm.exe` and `scumm.dll`.
`scumm.dll` is a DLL with the engine we selected and it uses `scummvm.exe` as a "library" for getting the functions it needs.
It is possible to verify it by using any `mingw-w64` cross compiler as I did.

Importing functions from an exe file like you do for DLLs works fine on Windows for desktops, but it doesn't on WinCE.
`LoadLibrary()` always fails on WinCE if you try to load an exe, it must be a DLL and it seems that there is no way to bypass it or at least I haven't found it.

In my opinion, the only solution is to modify the way plugins are made: there will be a small `scummvm.exe` file created from `backends\platform\sdl\win32\win32-main.cpp`, then a big `libscummvm.dll` that holds everything was into the original `scummvm.exe` except the main function and finally the `scumm.dll` engine that will link to `libscummvm.dll` instead of `scummvm.exe`.
This is probably the biggest fix to do here because it seems it cannot be done in a different manner.
Actually, it seems to me that all releases of SCUMMVM for Windows are "monolithic" gigantic executables and never used shared libraries, so I don't know if it was done the way it is now for some reasons.

Besides that, the changes to SCUMMVM are limited to some preprocessor tests to `_WIN32_WCE`, some additions to the autoconf script and .mk and a dedicated `backends\platform\sdl\win32\wince-main.cpp` for starting `scummvm_main()`.

Thank you very much for your time.

sev- commented:
You do not need to mess with win32 backend. There is nothing to "fix". Create another backed, specific for WinCE, perhaps, subclassing Win32 if you like. But when I see things like `backends\platform\sdl\win32\wince-main.cpp`, e.g. wince within win32, that cries for troubles.

Also, I highly suggest to check the work that you have learned is already there. Maybe you could reuse the old port or at least borrow parts from it.

In general, feel free to come to our Discord and have discussions over here. Having them in the closed ticket is a bit bizarre.
